### DISCUSSION TITLE
Replace IEqualityComparer with "==" interface

### DISCUSSION THREAD

**Discussion by YT-GameMaker** [dotnet/csharplang#7641] [Language Ideas]
```
class Object
{
    virtual bool Equals(object obj);
    static bool Equals(object objA, object objB);
}

struct Single
{
    bool operator ==(Single left, Single right);
}

interface IEquatable<T>
{
    bool Equals(T other);
}

interface IEqualityComparer<in T>
{
    bool Equals(T x, T y);
}

interface IComparable<in T>
{
    int CompareTo(T other);
}
```
I don't know how you feel, but every time I see these things, I feel suffocated. Checking whether two objects are equal, comparing two objects, these operations are unreasonably designed as ordinary members, which requires checking for null references when calling. At the same time, the semantic overlap between the "==" operator and ordinary methods is frustrating.

Now you can use the "==" interface to replace IEqualityComparer to unify semantics, and type T can directly implement this interface. IEquatable, IEqualityComparer, and Equals(object) can all be discarded. Note that the "==" interface can still be implemented by types other than T, and you only need to pass the comparison type you want to use when using it. Comparison can also be improved in a similar way.

**Comment by HaloFour** [comment]
I'm not sure what you mean by "replace".  It's not like the runtime can remove those interfaces, as too many other classes depend on them.  I'd suggest that `IEqualityOperators<TSelf,TOther,TResult>` would be more awkward to use, and doesn't cover as many use cases (e.g. doesn't calculate custom hash codes, which is necessary for dictionary scenarios).

**Reply by YT-GameMaker** [reply]
Custom hash codes can be implemented through a new interface with `static abstract GetHashCode`, just like `IEqualityOperators`.

**Comment by CyrusNajmabadi** [comment]
I'm not really understanding what the concrete suggestion is here at the language level. Could you flesh it out a bit?

**Reply by YT-GameMaker** [reply]
The main changes are not in the language level, but the language should allow `IEqualityOperators` to be implemented by any type (not just related types), then users can custom comparison types.

**Comment by tannergooding** [comment]
Please see the response provided at https://github.com/dotnet/runtime/discussions/94269#discussioncomment-7450890 for more context.

`IEqualityOperators` (`operator ==`) and `IEquatable` (`Equals`) are not equivalent and cannot simply be exchanged with each other.

**Comment by CyrusNajmabadi** [comment]
Closing out. This is one of many duplicate proposals from the same user.
